### PR TITLE
Revert non-best block

### PR DESCRIPTION
**Pull request by bkchr**
This makes `revert` also revert non-best blocks.

Closes: https://github.com/paritytech/cumulus/issues/1393

### CODE REVIEW

File: client/db/src/lib.rs [hunk 1]
```diff
@@ -2057,36 +2057,46 @@ impl<Block: BlockT> sc_client_api::backend::Backend<Block> for Backend<Block> {
...
+			.read()
+			.highest_leaf()
+			.and_then(|(n, h)| h.last().map(|h| (n, *h)));
+
+		let best_number = info.best_number;
+		let best_hash = info.best_hash;
+
+		let finalized = info.finalized_number;
+
+		let (mut number_to_revert, mut hash_to_revert) = match dbg!(highest_leaf) {
```
Comment (bkchr) [inline client/db/src/lib.rs:2074]:
```suggestion
		let (mut number_to_revert, mut hash_to_revert) = match highest_leaf {
```

**Comment by davxy** [comment]
If I understood correctly:

1. Instead of always begin reverting from the best block (lets call this *B*) with this modification we start reverting from one leaf *R* such that height(*R*) >= height(*B*)  (can be *B* itself if they have the same height).
2. Reverts *R*'s branch down to **at most** the height of *B* and then it eventually jumps into *B*'s branch, continuing the revert loop into this branch instead.
3. This is performed on only one branch *R*. If there are 2+ branches at the same height of *R* then only one is reverted (the last one in the leaves Vec). Leaving the others as is.
4. This is done also when the branch *R* starts below the revert limit height. Thus we may end up partially reverting branches create before the revert point.

If this is the intended behavior then ... OK! :-D 

But I've not completely understood  why in this case we limit to revert just one branch R of the multiple ones that are eventually available.

---

Beside this , I have one concern about the usage of the input parameter `n`

I try to explain with an example.
Let's assume `revert_finalized=false`, `n=1`, `best_num=3`, `finalized_num=1` and `highest_leaf_num=4`.

We set `revertible = highest_leaf_num - finalized_num = 3`.

According to https://github.com/paritytech/substrate/blob/aacc3e99c0750c3e6c851a8270dc2c4eebca3290/client/db/src/lib.rs#L2066

since `revertible < n` is false, we set `n = 1` (the original value).

But in the loop we start reverting from `highest_leaf`, so we revert only n=1 blocks and we stop at height 3,
while according to the previous implementation maybe the user expects a reversion of n=1 blocks from the best one, i.e. stop at block 2 (even though this is not explicitly documented in the `Backend` trait API).

This is not catastrophic, but then babe and grandpa reverts are going to revert 1 block from the best one. i.e. down to block 2.

Please correct if I missed something

An easy fix could be to set:

```
		let (mut number_to_revert, mut hash_to_revert) = match highest_leaf {
			Some((num_to_revert, hash_to_revert)) if num_to_revert > best_number => {
			    n = n + (num_to_revert - best_number); // tweak the original n parameter
			    (num_to_revert, hash_to_revert),
			}
			_ => (best_number, best_hash),
		};
```

**Comment by davxy** [comment]
And yeah... sorry for the long comment :-D

**Comment by bkchr** [comment]
> * Instead of always begin reverting from the best block (lets call this _B_) with this modification we start reverting from one leaf _R_ such that height(_R_) >= height(_B_)  (can be _B_ itself if they have the same height).
> 
> * Reverts _R_'s branch down to **at most** the height of _B_ and then it eventually jumps into _B_'s branch, continuing the revert loop into this branch instead.
> 
> * This is performed on only one branch _R_. If there are 2+ branches at the same height of _R_ then only one is reverted (the last one in the leaves Vec). Leaving the others as is.
> 
> * This is done also when the branch _R_ starts below the revert limit height. Thus we may end up partially reverting branches create before the revert point.

So, the current implementation is broken. If you for example have a leaf that has a height (3) above the best block (2) and the finalized block (1) and we are reverting one block, it would currently lead to reverting only the leaf. Because we call this `revert_one` function that just reverts whatever has the highest block number. Reverting on a certain height also always means we are reverting all blocks on this height, not just any block of the randomly chosen branch. I switch back to the branch when we are at the height of the best block, to set the correct best block based on the current best block branch.

**Comment by bkchr** [comment]
> This is not catastrophic, but then babe and grandpa reverts are going to revert 1 block from the best one. i.e. down to block 2.

This sounds to me like another bug/problem. I mean that would mean that there is maybe data to blocks associated that we don't revert. However, I can tweak the implementation to let `n` only take blocks into account from the best block downwards.

**Comment by davxy** [comment]
> This sounds to me like another bug/problem. I mean that would mean that there is maybe data to blocks associated that we don't revert. 

If you intend in babe/grandpa revert procedure, then all leaves branching **after** the revert point are correctly wiped out. 

The data that belongs to branches starting **below** the revert point is not touched.

This no dangling data is left around

**Comment by bkchr** [comment]
> > This sounds to me like another bug/problem. I mean that would mean that there is maybe data to blocks associated that we don't revert.
> 
> If you intend in babe/grandpa revert procedure, then all leaves branching **after** the revert point are correctly wiped out.
> 
> Only the data that belongs to branches starting **below** the revert point is not touched.
> 
> Thus no dangling data should be left around

What I meant is that Babe/Grandpa currently also only start with the best block and ignoring any leaves with higher block numbers.

**Comment by davxy** [comment]
> What I meant is that Babe/Grandpa currently also only start with the best block and ignoring any leaves with higher block numbers.

Maybe we are not talking about the same operation, but I'm saying that - on revert - even tough we are reverting from the best block, in babe/grandpa we are removing the data associated to ANY branch starting **after** the revert point (also the ones with higher numbers).

In babe: 
https://github.com/paritytech/substrate/blob/aacc3e99c0750c3e6c851a8270dc2c4eebca3290/client/consensus/babe/src/lib.rs#L1905-L1909

In grandpa:
https://github.com/paritytech/substrate/blob/aacc3e99c0750c3e6c851a8270dc2c4eebca3290/client/finality-grandpa/src/authorities.rs#L228-L248

**Comment by davxy** [comment]
Just to be sure that we are also aligned on the words :-) 

For **revert-point** I mean the point where you arrive **after** the revert operation (not the start, e.g the best block)

**Comment by bkchr** [comment]
> even tough we are reverting from the best block, in babe/grandpa we are removing the data associated to ANY branch starting **after** the revert point (also the ones with higher numbers).

Okay, yes, I really understood it differently before. However, there is still a difference to what the backend is doing. The backend is deleting all blocks that are on the same height, even if they don't originate from the revert point.

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -3463,4 +3481,71 @@ pub(crate) mod tests {
...
+		assert!(backend.have_state_at(&block4, 4));
+
+		assert_eq!(3, backend.revert(1, false).unwrap().0);
+
+		assert!(backend.have_state_at(&block1, 1));
+		assert!(!backend.have_state_at(&block2, 2));
+		assert!(!backend.have_state_at(&block3, 3));
+		assert!(!backend.have_state_at(&block4, 4));
+
+		assert_eq!(1, backend.blockchain.leaves.read().highest_leaf().unwrap().0);
```
Comment (davxy) [inline client/db/src/lib.rs:3549]:
```suggestion
		let block5 = insert_header_no_head(&backend, 3, block2, Default::default());

		assert!(backend.have_state_at(&block1, 1));
		assert!(backend.have_state_at(&block2, 2));
		assert!(backend.have_state_at(&block3, 3));
		assert!(backend.have_state_at(&block4, 4));
		assert!(backend.have_state_at(&block5, 3));

		assert_eq!(backend.blockchain.leaves().unwrap(), vec![block4, block5]);
		assert_eq!(4, backend.blockchain.leaves.read().highest_leaf().unwrap().0);

		assert_eq!(3, backend.revert(1, false).unwrap().0);

		assert!(backend.have_state_at(&block1, 1));
		assert!(!backend.have_state_at(&block2, 2));
		assert!(!backend.have_state_at(&block3, 3));
		assert!(!backend.have_state_at(&block4, 4));
		assert!(!backend.have_state_at(&block5, 3));

		assert_eq!(backend.blockchain.leaves().unwrap(), vec![block1]);
		assert_eq!(1, backend.blockchain.leaves.read().highest_leaf().unwrap().0);
```

**Review by davxy — APPROVED** [review]
LGTM

I've just modified the test assertions section to excercise the fact that the backend reversion is performed "per level" (as you've pointed out [here](https://github.com/paritytech/substrate/pull/11716/files#diff-0e523389a21c9bd9fc1da47816eb73a5ad21637337bb938abecedfe7f936d824R3537-R3549))

File: client/service/src/chain_ops/revert_chain.rs [hunk 3]
```diff
@@ -39,7 +39,14 @@ where
 	if reverted.0.is_zero() {
 		info!("There aren't any non-finalized blocks to revert.");
 	} else {
-		info!("Reverted {} blocks. Best: #{} ({})", reverted.0, info.best_number, info.best_hash);
+		info!("Reverted {} blocks. Best: #{} ({})", reverted.0, info.best_number, info.best_hash,);
```
Comment (davxy) [inline client/service/src/chain_ops/revert_chain.rs:42]:
```suggestion
		info!("Reverted {} blocks. Best: #{} ({})", reverted.0, info.best_number, info.best_hash);
```

petty nitpick

**Review by arkpar — APPROVED** [review]

**Comment by bkchr** [comment]
bot merge

**Pull request merged**